test/core/client_channel/lb_policy/xds_override_host_test.cc

```diff
@@ -86,6 +121,97 @@ TEST_F(XdsOverrideHostTest, NoConfigReportsError) {
       absl::InvalidArgumentError("Missing policy config"));
 }
 
+TEST_F(XdsOverrideHostTest, OverrideHost) {
+  // Send address list to LB policy.
+  const std::array<absl::string_view, 3> kAddresses = {
+      "ipv4:127.0.0.1:441", "ipv4:127.0.0.1:442", "ipv4:127.0.0.1:443"};
+  auto maybe_picker = InitRoundRobin(kAddresses);
+  ASSERT_TRUE(maybe_picker.has_value());
+  auto picker = *maybe_picker;
+  ASSERT_NE(picker, nullptr);
+  // Check that the host is overridden
+  std::map<UniqueTypeName, std::string> call_attributes{
+      {XdsHostOverrideTypeName(), "127.0.0.1:442"}};
+  EXPECT_EQ(ExpectPickComplete(picker.get(), call_attributes), kAddresses[1]);
+  EXPECT_EQ(ExpectPickComplete(picker.get(), call_attributes), kAddresses[1]);
+  call_attributes[XdsHostOverrideTypeName()] = std::string("127.0.0.1:441");
+  EXPECT_EQ(ExpectPickComplete(picker.get(), call_attributes), kAddresses[0]);
+  EXPECT_EQ(ExpectPickComplete(picker.get(), call_attributes), kAddresses[0]);
+}
+
+TEST_F(XdsOverrideHostTest, OverrideHostSubchannelNotFound) {
+  // Send address list to LB policy.
+  const std::array<absl::string_view, 3> kAddresses = {
+      "ipv4:127.0.0.1:441", "ipv4:127.0.0.1:442", "ipv4:127.0.0.1:443"};
+  auto maybe_picker = InitRoundRobin(kAddresses);
+  ASSERT_TRUE(maybe_picker.has_value());
+  auto picker = *maybe_picker;
+  ASSERT_NE(picker, nullptr);
+  std::map<UniqueTypeName, std::string> call_attributes{
+      {XdsHostOverrideTypeName(), "no such host"}};
+  std::unordered_set<std::string> picks;
+  for (size_t i = 0; i < kAddresses.size(); ++i) {
+    auto pick = ExpectPickComplete(picker.get(), call_attributes);
+    ASSERT_TRUE(pick.has_value());
+    picks.emplace(*pick);
+  }
+  ASSERT_THAT(picks, UnorderedElementsAreArray(kAddresses));
+}
+
+TEST_F(XdsOverrideHostTest, SubchannelsComeAndGo) {
+  const std::array<absl::string_view, 3> kAddresses = {
+      "ipv4:127.0.0.1:441", "ipv4:127.0.0.1:442", "ipv4:127.0.0.1:443"};
+  auto maybe_picker = InitRoundRobin(kAddresses);
+  ASSERT_TRUE(maybe_picker.has_value());
+  auto picker = *maybe_picker;
+  ASSERT_NE(picker, nullptr);
+  // Make sure child policy works
+  EXPECT_NE(ExpectPickComplete(picker.get()), ExpectPickComplete(picker.get()));
+  // Check that the host is overridden
+  std::map<UniqueTypeName, std::string> call_attributes{
+      {XdsHostOverrideTypeName(), "127.0.0.1:442"}};
+  EXPECT_EQ(ExpectPickComplete(picker.get(), call_attributes), kAddresses[1]);
+  EXPECT_EQ(ExpectPickComplete(picker.get(), call_attributes), kAddresses[1]);
+  // Some other address is gone
+  EXPECT_EQ(
+      ApplyUpdate(BuildUpdate({"ipv4:127.0.0.1:441", "ipv4:127.0.0.1:442"},
+                              MakeXdsOverrideHostConfig("round_robin")),
+                  policy_.get()),
+      absl::OkStatus());
+  ExpectState(GRPC_CHANNEL_READY);
```

Comment:
This is still (a) checking for individual state changes instead of checking only when we actually expect a behavior change and (b) getting a state change and not checking the resulting picker behavior.

I think we should instead use `WaitForRoundRobinListChange()` here to wait for the LB policy to switch to the new address list.

Reply:
Why do you need `ExpectState(GRPC_CHANNEL_READY)` here?  I don't think that should be necessary.  What error do you get if you remove that line?

I don't think tests should have to directly call `ExpectState()` directly in the vast majority of cases.  And if there is some edge case in which a test does need to do that, it cannot ignore the returned picker -- tests must *always* verify the behavior of the picker.

Reply:
It looks like there were two problems here.

1. There was a bug in RR causing it to generate spurious picker updates whenever it swapped out the subchannel list.  I've fixed this in #31939.
2. The purpose of calling `WaitForRoundRobinListChange()` here is to ensure that the LB policy has actually swapped over to the new address list.  You can't check that by using the call attributes here, because then you're forcing the picker to return a specific address instead of checking that it's returning the new list of addresses that we're looking for.  Instead, you need to first call `WaitForRoundRobinListChange()` to check for the new list without the call attributes, and then use the call attributes to make sure they're working as expected.  I've made this change in https://github.com/eugeneo/grpc/pull/1.